File: 2900/c/2900.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

 char** getLongestSubsequence(char** words, int wordsSize, int* groups, int groupsSize, int* returnSize) {
    // make dp, preWord
    int* dp = (int*)malloc(wordsSize * sizeof(int));
    int* preWord = (int*)malloc(wordsSize * sizeof(int));
    // init
    for (int i = 0; i < wordsSize; i++){
        dp[i] = 1;
        preWord[i] = -1;
    }

    // state transition equation
    for (int i = 1; i < wordsSize; i++){
        for (int j = 0; j < i; j++){
            if (groups[i] != groups[j]){
                if (dp[i] < dp[j] + 1){
                    dp[i] = dp[j] + 1;
                    preWord[i] = j;
                }
            }
        }
    }

    // find the longest
    int max = 0, lastId = -1;
    for (int i = 0; i < wordsSize; i++){
        if (max < dp[i]){
            max = dp[i];
            lastId = i;
        }
    }
    // form the result from preWord
    char** result = (char**)malloc(max * sizeof(char*));
    for (int i = max-1; i >= 0; i--){
        result[i] = words[lastId];
        lastId = preWord[lastId];
    }

    free(dp);
    free(preWord);

    *returnSize = max;

    return result;
}

File: 2900/py/2900.py
class Solution:
    def getLongestSubsequence(self, words: List[str], groups: List[int]) -> List[str]:
        n = len(words)
        # make dp
        dp = [1 for _ in range(n)]
        # make recorder
        preWord = [-1 for _ in range(n)]

        # state transition
        for i in range(1, n) :
            for j in range(0, i) :
                if groups[i] != groups[j] :
                    if dp[i] < dp[j]+1 :
                        dp[i] = dp[j]+1
                        preWord[i] = j
        
        # get the longest
        maxLen = 0
        endId = -1
        for i in range(n) :
            if maxLen < dp[i] :
                maxLen = dp[i]
                endId = i
        
        # form the result
        result = ["" for _ in range(maxLen)]

        for i in range(maxLen-1, -1, -1) :
            result[i] = words[endId]
            endId = preWord[endId]

        return result